instrumentation/kafka/src/main/java/brave/kafka/TracingProducer.java

```diff
@@ -0,0 +1,118 @@
+package brave.kafka;
+
+import brave.Span;
+import brave.Span.Kind;
+import brave.Tracing;
+import brave.propagation.TraceContext;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.ProducerFencedException;
+import org.apache.kafka.common.header.Headers;
+import zipkin.internal.Util;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+public class TracingProducer<K, V> implements Producer<K, V> {
+
+    public static final String KAFKA_TOPIC_TAG = "kafka.topic";
+    public static final String KAFKA_KEY_TAG = "kafka.key";
+
+    private Tracing tracing;
+    private TraceContext.Injector<Headers> injector;
+    private Producer<K, V> wrappedProducer;
+
+    public TracingProducer(Tracing tracing, Producer<K, V> producer) {
+        this.wrappedProducer = producer;
+
+        this.tracing = tracing;
+        this.injector = tracing.propagation().injector(
+                (carrier, key, value) -> carrier.add(key, value.getBytes(Util.UTF_8)));
+    }
+
+    @Override
+    public void initTransactions() {
+        wrappedProducer.initTransactions();
+    }
+
+    @Override
+    public void beginTransaction() throws ProducerFencedException {
+        wrappedProducer.beginTransaction();
+    }
+
+    @Override
+    public void commitTransaction() throws ProducerFencedException {
+        wrappedProducer.commitTransaction();
+    }
+
+    @Override
+    public void abortTransaction() throws ProducerFencedException {
+        wrappedProducer.abortTransaction();
+    }
+
+    /**
+     * Send with a callback is always called for KafkaProducer.
+     * We do the same here to enable tracing.
+     */
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+        return this.send(record, null);
+    }
+
+    /**
+     * We wrap the send method to add tracing.
+     */
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+        Span span = tracing.tracer().nextSpan();
+        span.tag(KAFKA_TOPIC_TAG, record.topic());
+        span.tag(KAFKA_KEY_TAG, record.key().toString());
```

Comment:
key might be null? also maybe just start with topic? or get input from others before doing key.